staging/src/k8s.io/client-go/plugin/pkg/client/auth/azure/azure.go

```diff
@@ -180,34 +180,49 @@ func (ts *azureTokenSource) Token() (*azureToken, error) {
 	defer ts.lock.Unlock()
 
 	var err error
+	// 1. use cache if available
 	token := ts.cache.getToken(azureTokenKey)
+
+	// 2. load persist if no cache
 	if token == nil {
 		token, err = ts.retrieveTokenFromCfg()
-		if err != nil {
-			token, err = ts.source.Token()
-			if err != nil {
-				return nil, fmt.Errorf("acquiring a new fresh token: %v", err)
-			}
-		}
-		if !token.token.IsExpired() {
+
+		// set cfg to cache 
+		if err == nil {
+			klog.V(4).Infof("Saving cache after load cfg")
 			ts.cache.setToken(azureTokenKey, token)
-			err = ts.storeTokenInCfg(token)
-			if err != nil {
-				return nil, fmt.Errorf("storing the token in configuration: %v", err)
-			}
 		}
 	}
-	if token.token.IsExpired() {
+	
+	// quick exit for caching scenario
+	if token != nil && !token.token.IsExpired() {
+		return token, nil
+	}
+
+	// 3. extend if valid token but expired
+	if token != nil && token.token.IsExpired() {
+		klog.V(4).Infof("Refreshing token.")
 		token, err = ts.refreshToken(token)
 		if err != nil {
```

Comment:
can we add a comment here:
// when refreshToken fails, token will be reset to nil so that the inner token source will be used to acquire new token